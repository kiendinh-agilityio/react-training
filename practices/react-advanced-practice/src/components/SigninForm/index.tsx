// Import react hook form
import { useForm, SubmitHandler, Controller } from 'react-hook-form';

// Import radix ui
import { Box } from '@radix-ui/themes';

// Import common components
import { Input, Button, Text } from '@/components/common';

// Import constants
import { REGEX, MESSAGE_ERROR } from '@/constants';

// Import icons
import { ShowPasswordIcon, HidePasswordIcon } from '@/components/common/Icons';

// Import stores
import { useSigninStore } from '@/stores';

// Define form input types
interface SigninFormInputs {
  email: string;
  password: string;
}

const SigninForm = () => {
  const { showPassword, togglePasswordVisibility, setEmail, setPassword } =
    useSigninStore();

  const {
    control,
    trigger,
    handleSubmit,
    formState: { errors },
  } = useForm<SigninFormInputs>();

  const onSubmit: SubmitHandler<SigninFormInputs> = (data) => {
    setEmail(data.email);
    setPassword(data.password);
  };

  /**
   * Creates an event handler for the blur event to validate a specific field.
   * @param fieldName - The name of the field to validate.
   * @param trigger - The react-hook-form trigger function.
   * @returns A function that triggers validation for the specified field.
   */
  const handleFieldBlur =
    (
      fieldName: keyof SigninFormInputs,
      trigger: (fieldName: keyof SigninFormInputs) => void,
    ) =>
    () =>
      trigger(fieldName);

  /**
   * Handles the onChange event for form inputs and triggers validation.
   * @param fieldName - The name of the field to validate.
   * @param fieldOnChange - The react-hook-form onChange handler for the field.
   * @returns A function that handles the input's onChange event.
   */
  const handleFieldChange =
    (fieldName: keyof SigninFormInputs, fieldOnChange: (value: string) => void) =>
    (e: React.ChangeEvent<HTMLInputElement>) => {
      fieldOnChange(e.target.value);
      trigger(fieldName);
    };

  return (
    <form onSubmit={handleSubmit(onSubmit)} className="flex flex-col max-w-[350px]">
      {/* Email Input */}
      <Box className="flex flex-col mb-6 gap-[5px]">
        <Text content="Email" />
        <Controller
          name="email"
          control={control}
          rules={{
            required: MESSAGE_ERROR.REQUIRED_ERROR('Email'),
            pattern: {
              value: REGEX.EMAIL,
              message: MESSAGE_ERROR.INVALID_EMAIL,
            },
          }}
          render={({ field }) => (
            <Input
              {...field}
              placeholder="Your email address"
              type="email"
              errorMessage={errors.email?.message}
              onChange={handleFieldChange('email', field.onChange)}
              onBlur={handleFieldBlur('email', trigger)}
            />
          )}
        />
      </Box>

      {/* Password Input */}
      <Box className="flex flex-col gap-[5px]">
        <Text content="Password" />
        <Controller
          name="password"
          control={control}
          rules={{
            required: MESSAGE_ERROR.REQUIRED_ERROR('Password'),
            pattern: {
              value: REGEX.PASSWORD,
              message: MESSAGE_ERROR.INVALID_PASSWORD,
            },
          }}
          render={({ field }) => (
            <Input
              {...field}
              placeholder="Your password"
              type={showPassword ? 'text' : 'password'}
              errorMessage={errors.password?.message}
              onChange={handleFieldChange('password', field.onChange)}
              onBlur={handleFieldBlur('password', trigger)}
              rightIcon={
                <button
                  type="button"
                  onClick={togglePasswordVisibility}
                  aria-label={showPassword ? 'Hide password' : 'Show password'}
                >
                  {showPassword ? <ShowPasswordIcon /> : <HidePasswordIcon />}
                </button>
              }
            />
          )}
        />
      </Box>

      {/* Submit Button */}
      <Button onClick={handleSubmit(onSubmit)} className="mt-9 w-[350px]">
        Sign In
      </Button>
    </form>
  );
};

export default SigninForm;
