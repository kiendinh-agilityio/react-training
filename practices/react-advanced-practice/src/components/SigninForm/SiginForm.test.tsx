import { render } from '@testing-library/react';
import SigninForm from '.';

describe('SigninForm Component', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<SigninForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
