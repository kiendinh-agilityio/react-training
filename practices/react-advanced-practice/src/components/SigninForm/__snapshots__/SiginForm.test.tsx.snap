// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SigninForm Component should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex flex-col max-w-[350px]"
  >
    <div
      class="rt-Box flex flex-col mb-6 gap-[5px]"
    >
      <p
        class="rt-Text rt-r-weight-regular text-sm "
      >
        Email
      </p>
      <div
        class="w-full"
      >
        <div
          class="rt-Flex rt-r-ai-center rt-r-jc-center flex justify-center items-center px-5 py-[15px] border gap-3 w-full rounded-[15px] border-input"
        >
          <input
            class="w-full text-regular focus:outline-none text-dark placeholder:text-base"
            name="email"
            placeholder="Your email address"
            type="email"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="rt-Box flex flex-col gap-[5px]"
    >
      <p
        class="rt-Text rt-r-weight-regular text-sm "
      >
        Password
      </p>
      <div
        class="w-full"
      >
        <div
          class="rt-Flex rt-r-ai-center rt-r-jc-center flex justify-center items-center px-5 py-[15px] border gap-3 w-full rounded-[15px] border-input"
        >
          <input
            class="w-full text-regular focus:outline-none text-dark placeholder:text-base"
            name="password"
            placeholder="Your password"
            type="password"
            value=""
          />
          <button
            aria-label="Show password"
            type="button"
          >
            <svg
              fill="none"
              height="23"
              viewBox="0 0 28 28"
              width="23"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                clip-rule="evenodd"
                fill="#a0aec0"
                fill-rule="evenodd"
              >
                <path
                  d="M17.747 15.415c.238-.54.372-1.143.372-1.413 0-2.348-1.824-4.25-4.073-4.25-2.25 0-4.073 1.902-4.073 4.25 0 2.347 2.037 3.887 4.073 3.887 1.303 0 2.462-.637 3.206-1.63a4.29 4.29 0 0 0 .495-.844Zm-3.701.349c.51 0 .975-.196 1.332-.518.432-.39.622-.965.622-1.244 0-1.174-.828-2.126-1.954-2.126-1.125 0-2.046.95-2.046 2.124 0 .81.92 1.764 2.046 1.764Z"
                />
                <path
                  d="M1.092 14.272a.448.448 0 0 1 0-.545c.01-.012.019-.025.028-.039C4.78 8.343 9.32 5.5 14.014 5.5c4.689 0 9.224 2.836 12.882 8.17.069.101.104.205.104.33 0 .13-.04.24-.114.346C23.23 19.67 18.698 22.5 14.014 22.5c-4.705 0-9.257-2.857-12.922-8.228Zm2.847-.92a1.04 1.04 0 0 0 0 1.294c3.135 3.898 6.665 5.729 10.075 5.729 3.41 0 6.94-1.83 10.075-5.729a1.04 1.04 0 0 0 0-1.294c-3.135-3.897-6.665-5.727-10.075-5.727-3.41 0-6.94 1.83-10.075 5.727Z"
                />
              </g>
            </svg>
          </button>
        </div>
      </div>
    </div>
    <button
      class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button  
    px-[22px] py-2.5 font-bold rounded-xl cursor-pointer primary
   mt-9 w-[350px]"
      data-accent-color=""
    >
      Sign In
    </button>
  </form>
</DocumentFragment>
`;
